### PR TITLE
Deduplicate student assessment instance router

### PR DESCRIPTION
**Pull request by jonatanschroeder**
Continuation of #9108, but for assessment instances. ~~Pending: instance question table.~~

### CODE REVIEW

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js [hunk 1]
```diff
@@ -9,14 +9,27 @@ const error = require('@prairielearn/error');
...
+
+  // we updated the assessment_instance, so reload it
+  const result = sqldb.queryRow(
+    sql.select_assessment_instance,
+    { assessment_instance_id: locals.assessment_instance.id },
+    AssessmentInstanceSchema,
+  );
+  locals.assessment_instance = result.rows[0];
+}
+
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js:29]: Since this had to be added to exam I took the liberty of modernizing it.

Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js]: I had to revert the `queryRow` schema since `AssessmentInstanceSchema` is not declared yet. I want to minimize new features here, so I'm reverting to `queryAsync` for now.

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js [hunk 2]
```diff
@@ -161,6 +172,14 @@ router.get(
...
         res.locals.rolesInfo = groupInfo.rolesInfo;
         res.locals.used_join_code = req.body.used_join_code;
+
+        if (groupConfig.has_roles) {
+          const result = await groupAssessmentHelper.getAssessmentPermissions(
+            res.locals.assessment.id,
+            res.locals.user.user_id,
+          );
+          res.locals.canViewRoleTable = result.can_assign_roles_at_start;
+        }
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js:182]: Was it a bug that this wasn't here before?

**Review by nwalters512 — COMMENTED** [review]
I admit to not looking super closely at `apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs`; it was kind of hard to parse the diff. I'd appreciate recommendations on how to prove to myself that this is right (e.g. whatever you did locally when working on this change).

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js (hunk 2 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js:182]: Yes, exams did not have the role table in place.

**Comment by jonatanschroeder** [comment]
> I admit to not looking super closely at `apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs`; it was kind of hard to parse the diff. I'd appreciate recommendations on how to prove to myself that this is right (e.g. whatever you did locally when working on this change).

The way I did the diff was in a terminal with `diff -y -W200 -d -w` (adjust the 200 to the width of your terminal/monitor). It is not an easy pair of files to compare, though. That said, you should see that there are large blocks (particularly in the table rows, in the modals at the start, and in the footer) that have specific checks for assessment type. There are also some parts of the code that make the assumption that, if the type is homework, then allow_real_time_grading will be true by default, so is allow_real_time_grading is false then you can assume it's an Exam.

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs [hunk 3]
```diff
@@ -179,9 +184,16 @@
             <% } else { %>
             <tr>
               <th>Question</th>
-              <th class="text-center">Submission status</th>
+              <% if (assessment.type === 'Exam') { %><th class="text-center">Submission status</th><% } %>
             <% } %>
-            <% if (has_auto_grading_question && assessment.allow_real_time_grading) { %>
+
+            <% if (assessment.type === 'Homework') { %>
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs:190]: To make this all easier to understand and modify, I'd recommend mirroring the order and structure of conditionals between the `<thead>` and `<tbody>` sections as closely as possible (if it's indeed possible to do that in a sensible way - it feels like it should be, but I'm not 100% sure if this suggestion would pan out). That's the thing I'm currently struggling the most with when reviewing.

Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs:190]: tbody already has that. Note that other than the question zone titles and question label, all other columns are already behind a big if/else on the assessment type. I'll try to rearrange the thead to make it a bit easier to follow.

Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs:205]: Hopefully the changes in 27c8d95c8 should make it easier to follow?

Comment (nwalters512) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs:205]: This is better, thanks. To be more explicit with my suggestion, when I say "mirror the order and structure of conditionals", I'm referring to the fact that the `<thead>` leads with `if (assessment.type === 'Homework')`, whereas `<tbody>` leads with `if (assessment.type === 'Exam')`. IMO it'd be easier to mentally parse this page if e.g. both used `if (assessment.type === 'Homework') { ... } else { ... }`. Another difference: `<thead>` uses two `if (assessment.type === 'Homework') { ... } else { ... }` blocks back-to-back, whereas `<tbody>` uses just a single one, which makes it easier to parse what's happening in each case.

Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs:205]: I can rearrange the order. As for the two blocks back-to-back, they refer to similar functionality, so I'd prefer to keep them separate if you don't mind.

Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs]: Reordered in 5ebbebdc2.

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs [hunk 4]
```diff
@@ -167,47 +172,71 @@
...
+                  <%- include('../partials/examQuestionHelpAwardedPoints'); %>
+                </th>
+              <% } else if (has_auto_grading_question && has_manual_grading_question) { %>
+                <th class="text-center">Auto-grading points</th>
+                <th class="text-center">Manual grading points</th>
+                <th class="text-center">Total points</th>
+              <% } else { %>
+                <th class="text-center">Points</th>
+              <% } %>
+              <% } else { %>
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs:226]: Nit, to make this easier to parse.

```suggestion
            <% } else { %>
```

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs (hunk 3 above)
Comment (nwalters512) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs]: Thanks, I think I see now why the two sections can't perfectly mirror each other. It's because the header actually has multiple `<tr>` elements with `colspan` affecting the layout, whereas the body just has columns straight across, right?

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql [hunk 5]
```diff
@@ -1,3 +1,11 @@
+-- BLOCK select_assessment_instance
+SELECT
+  ai.*
+FROM
+  assessment_instances AS ai
+WHERE
+  ai.id = $assessment_instance_id;
+
 -- BLOCK select_instance_questions
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql:9]: Nit: there's a change in how `prev_title` is computed. Before, it would match what's displayed in the question list (e.g. "Question 2" for exams, "WS2.1" for homeworks). Now, it would show "Question 2" and "Question WS2.1". This is very much a nit, but capital-Q "Question WS2.1" feels off to me. I'm not yet requesting changes, just raising this for discussion.

**Review by nwalters512 — APPROVED** [review]

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql (hunk 5 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql:9]: Good catch. I'll add it to the batch. From what I could get, this is for forced order help text.

Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql:9]: Changed in 34cb0f1f1. There was no reason for "Question" to be part of the query, I moved it to the partial.

Comment (nwalters512) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql:9]: Note that "Question ###" (with a capital letter) is how exam questions are actually identified:

Screenshot 2024-01-03 at 15 07 20

I'm sure students will be able to tell that "question 1" and "Question 1" are the same thing 🙃 but the two ways that `prev_title` was previously being computed were very intentional.

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs (hunk 3 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.ejs]: Yes, that's correct. It just so happens (by coincidence) that the cell that uses colspan has 3 columns on both exams and homeworks, but the 3 columns are different, so keeping them separate probably makes sense to help with that case.

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql (hunk 5 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.sql:9]: I have changed it to have the same behaviour as before this PR. Changes can be discussed in the future.
